### PR TITLE
[Transforms] Let amdgcn take advantage of sin(-x) --> -sin(x)

### PR DESCRIPTION
**Pull request by AZero13**
We do it for amdgcn_cos, and we should do it for amdgcn_sin as well.

### CODE REVIEW

**Review by arsenm — CHANGES_REQUESTED** [review]
Missing test

**Comment by AtariDreams** [comment]
> Missing test

Fixed! @arsenm

File: llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll [hunk 1]
```diff
@@ -1023,6 +1023,50 @@ define float @cos_fabs_unary_fneg_f32(float %x) {
   ret float %cos
 }
 
+
+; --------------------------------------------------------------------
+; llvm.amdgcn.sin
+; --------------------------------------------------------------------
+declare float @llvm.amdgcn.sin.f32(float) nounwind readnone
+
+; CHECK-NEXT: ret float %cos
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:1032]: stray / broken check line?

File: llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll [hunk 2]
```diff
@@ -1023,6 +1023,50 @@ define float @cos_fabs_unary_fneg_f32(float %x) {
...
+define float @sin_fneg_f32(float %x) {
+; CHECK-LABEL: @sin_fneg_f32(
+; CHECK-NEXT:    [[TMP1:%.*]] = call float @llvm.amdgcn.sin.f32(float [[X:%.*]])
+; CHECK-NEXT:    [[SIN:%.*]] = fneg float [[TMP1]]
+; CHECK-NEXT:    ret float [[SIN]]
+;
+  %x.fneg = fsub float -0.0, %x
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:1039]: Prefer modern fneg instruction? Testing fsub -0, x may be interesting, but should test a true fneg

**Review by arsenm — APPROVED** [review]
This is the opposite of what we want for codegen, but I believe we undo this already

File: llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll (hunk 2 above)
Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:1039]: Other test has that.

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:1039]: What do you mean other test? I don't see another test? fneg is still preferable to magic fsub

**Review by arsenm — APPROVED** [review]

**Comment by yashssh** [comment]
Out of curiosity what's the purpose of doing this? From one of the tests attached It just looks like we are moving `fneg` from before `sin` to after.

**Comment by arsenm** [comment]
Why close this?

**Comment by AtariDreams** [comment]
@arsenm can we please merge this

**Comment by AtariDreams** [comment]
@arsenm Can we merge this now please?

File: llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll [hunk 3]
```diff
@@ -1023,6 +1023,47 @@ define float @cos_fabs_unary_fneg_f32(float %x) {
...
+; --------------------------------------------------------------------
+declare float @llvm.amdgcn.sin.f32(float) nounwind readnone
+
+define float @sin_fneg_f32(float %x) {
+; CHECK-LABEL: @sin_fneg_f32(
+; CHECK-NEXT:    [[X_FNEG:%.*]] = fsub float 0.000000e+00, [[X:%.*]]
+; CHECK-NEXT:    [[SIN:%.*]] = call float @llvm.amdgcn.sin.f32(float [[X_FNEG]])
+; CHECK-NEXT:    ret float [[SIN]]
+;
+  %x.fneg = fsub float 0.0, %x
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:1038]:
```suggestion
  %x.fneg = fneg float %x
```

File: llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll [hunk 4]
```diff
@@ -1023,6 +1023,47 @@ define float @cos_fabs_unary_fneg_f32(float %x) {
...
+
+define float @sin_fabs_fneg_f32(float %x) {
+; CHECK-LABEL: @sin_fabs_fneg_f32(
+; CHECK-NEXT:    [[X_FABS:%.*]] = call float @llvm.fabs.f32(float [[X:%.*]])
+; CHECK-NEXT:    [[TMP1:%.*]] = call float @llvm.amdgcn.sin.f32(float [[X_FABS]])
+; CHECK-NEXT:    [[SIN:%.*]] = fneg float [[TMP1]]
+; CHECK-NEXT:    ret float [[SIN]]
+;
+  %x.fabs = call float @llvm.fabs.f32(float %x)
+  %x.fabs.fneg = fsub float -0.0, %x.fabs
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:1062]:
```suggestion
  %x.fabs.fneg = fneg float %x.fabs
```

File: llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll [hunk 5]
```diff
@@ -1023,6 +1023,47 @@ define float @cos_fabs_unary_fneg_f32(float %x) {
...
+  %sin = call float @llvm.amdgcn.sin.f32(float %x.fabs.fneg)
+  ret float %sin
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:1066]: Add a test with flag preservation

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/AMDGPU/amdgcn-intrinsics.ll:1079]: Done!

**Pull request merged**